### DISCUSSION TITLE
Why doesn't git move do anything in worktrees?

### DISCUSSION THREAD

**Discussion by miquella** [arxanas/git-branchless#928] [General]
Hello wonderful friends!

I've been quite confused by `git move` for a long time and believe I realized why today: worktrees!

## Background

I tend to have a lot of incomplete streams of work that I switch between often, so I have several worktrees open at any given time (although `git-branchless` has helped to reduce this quite a bit, I haven't completely kicked the habit just yet 😜). To make matters worse, I tend to not use branches either.

While working in one of these branchless worktrees, I often come across an unrelated issue. So I fix the issue and commit it separate from the rest of my work. Once I reach a stopping point, I cherry-pick those changes onto the main branch and rebase them out of my commit history.

## Confusion

Once I picked up `git-branchless`, I realized that I could simplify (or so I thought…) this workflow by using `git move`!

The thing that's always confused me is that it never seemed to work. I would `git move -x <COMMIT> -d <MAIN>` and … nothing would change. Or worse, sometimes the commit would just be gone! Thankfully, when it was missing, some combination of `git undo` and/or scouring the output would allow me to recover the missing commit.

Today while trying to figure out what was going on, I realized this only fails in worktrees 😮 If I switched to my standard clone, everything works fine!

So my question is: am I doing something wrong? Perhaps I'm just expecting the wrong thing?

## Demo

To demonstrate the issue, I setup a new repo:

```console
$ mkdir branchless-demo
$ cd branchless-demo
$ git init
$ git branchless init --main-branch master

$ printf 'A\nB\nD\n' > README.md
$ git add README.md
$ git commit -m 'README'
[master (root-commit) e7d9765] README
 1 file changed, 3 insertions(+)
 create mode 100644 README.md
```

Switched to a worktree:

```console
$ git worktree add --detach ../branchless-demo-2
Preparing worktree (detached HEAD e7d9765)
branchless: processing 1 update: ref HEAD
branchless: processing 1 update: ref HEAD
HEAD is now at e7d9765 Something else
branchless: processing checkout
$ cd ../branchless-demo-2
```

And did some "work":

```console
$ touch something.txt
$ git add something.txt
$ git commit -m 'Something'
branchless: processing 1 update: ref HEAD
branchless: processed commit: 12e5739 Something
[detached HEAD 12e5739] Something
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 something.txt

$ printf 'A\nB\nC\nD\n' > README.md
$ git add README.md
$ git commit -m 'Update README'
branchless: processing 1 update: ref HEAD
branchless: processed commit: 0adb2e7 Update README
[detached HEAD 0adb2e7] Update README
 1 file changed, 1 insertion(+)

$ touch something_else.txt
$ git add something_else.txt
$ git commit -m 'Something else'
branchless: processing 1 update: ref HEAD
branchless: processed commit: bf05214 Something else
[detached HEAD bf05214] Something else
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 something_else.txt

$ git sl
◇ e7d9765 1m (master) README
┃
◯ 12e5739 32s Something
┃
◯ 0adb2e7 12s Update README
┃
● bf05214 2s Something else
```

Then attempt to move something:

```console
$ git move -x 0adb2e7 -d e7d9765
Attempting rebase in-memory...
[1/2] Committed as: d0dc1ac Something else
[2/2] Committed as: 58762a1 Update README
branchless: creating working copy snapshot
branchless: running command: git checkout d0dc1ace840fd1715329e8a15a094ddc2558ce2c
Previous HEAD position was bf05214 Something else
branchless: processing 1 update: ref HEAD
HEAD is now at d0dc1ac Something else
branchless: processing checkout
◇ e7d9765 41m (master) README
┃
◯ 12e5739 40m Something
┃
● d0dc1ac 0s Something else
In-memory rebase succeeded.
```

But now the "Update README" commit is gone from the smartlog! I've learned that if I carefully read the output, I can unhide the commit and then the smartlog appears as I would expect:

```console
$ git unhide 58762a1
Unhid commit: 58762a1 Update README
To hide this 1 commit, run: git undo
$ git sl
◇ e7d9765 44m (master) README
┣━┓
┃ ◯ 12e5739 43m Something
┃ ┃
┃ ● d0dc1ac 2m Something else
┃
◯ 58762a1 2m Update README
```

**Comment by miquella** [comment]
It does make me wonder if this is somehow related to #215?

**Comment by arxanas** [comment]
I haven't taken a look, but you might need to run `git branchless init` in the worktrees to install hooks.

**Comment by miquella** [comment]
Good thought!

Running `git branchless init` in the worktree doesn't appear to have changed anything:

```console
$ cd branchless-demo-2
$ git branchless init
Created config file at /Users/miquella/src/branchless-demo/.git/branchless/config
Auto-detected your main branch as: master
If this is incorrect, run: git branchless init --main-branch <branch>
Installing hook: post-commit
Installing hook: post-merge
Installing hook: post-rewrite
Installing hook: post-checkout
Installing hook: pre-auto-gc
Installing hook: reference-transaction
Successfully installed git-branchless.
To uninstall, run: git branchless init --uninstall

$ git sl
◇ e7d9765 6d (master) README
┃
◯ 12e5739 6d Something
┃
◯ 0adb2e7 6d Update README
┃
● bf05214 6d Something else

$ git move -x 0adb2e7 -d e7d9765
Attempting rebase in-memory...
[1/2] Committed as: c04d12b Something else
[2/2] Committed as: 0ef1c57 Update README
branchless: creating working copy snapshot
branchless: running command: git checkout c04d12b33f99e9c876826d5d6844946803357a78
Previous HEAD position was bf05214 Something else
branchless: processing 1 update: ref HEAD
HEAD is now at c04d12b Something else
branchless: processing checkout
◇ e7d9765 6d (master) README
┃
◯ 12e5739 6d Something
┃
● c04d12b 1s Something else
In-memory rebase succeeded.
```

The resulting commit (`0ef1c57`) still needed to be passed into `git unhide` for it to show up in the smartlog:

```console
$ git unhide 0ef1c57
Unhid commit: 0ef1c57 Update README
To hide this 1 commit, run: git undo

$ git sl
◇ e7d9765 6d (master) README
┣━┓
┃ ◯ 12e5739 6d Something
┃ ┃
┃ ● c04d12b 2m Something else
┃
◯ 0ef1c57 2m Update README
```

Correlating this with the [repository layout docs](https://git-scm.com/docs/gitrepository-layout#Documentation/gitrepository-layout.txt-hooks) seems to confirm that worktrees do share hooks with the original repo (although I'm not sure if `git-branchless` is doing much outside of the hooks).

**Comment by arxanas** [comment]
A colleague of mine reported problems with relative worktrees. Can you try with an absolute path to the worktree and see if that resolves the issue?

**Comment by miquella** [comment]
Oh, great idea!

Unfortunately, it looks like it's already an absolute path:
```console
$ cat .git
gitdir: /Users/miquella/src/branchless-demo/.git/worktrees/branchless-demo-2
```

**Comment by arxanas** [comment]
In my case, `git commit` indicates that it is invoking git-branchless hooks. That is, instead of

```
$ git commit -m 'README'
[master (root-commit) e7d9765] README
 1 file changed, 3 insertions(+)
 create mode 100644 README.md
 ```
 
 I get
 
 ```
$ git commit -m 'README'
branchless: processing 2 updates: branch master, ref HEAD
branchless: processed commit: 94fa741 README
[master (root-commit) 94fa741] README
 1 file changed, 3 insertions(+)
 create mode 100644 README.md
 ```
 
 Although later I see that your reference-transaction and post-commit hooks seem to have triggered properly.
 
After following along with the rest of the steps, it didn't reproduce, and the `git move` invocation immediately showed the new commits:
 
 ```
 ﻿﻿$ git sl
◇ 94fa741 3m (master) README
┃
◯ f4e2e91 2m Something
┃
◯ e31bf9c 1m Update README
┃
● 4856148 1m Something else
$ git move -s e31bf9c -d 94fa741
Attempting rebase in-memory...
[1/2] Committed as: 81816d2 Update README
[2/2] Committed as: 7118d31 Something else
branchless: creating working copy snapshot
branchless: running command: git checkout 7118d31fc9dbcc989d543b2a2426f0e1cce80bbc
Previous HEAD position was 4856148 Something else
branchless: processing 1 update: ref HEAD
HEAD is now at 7118d31 Something else
branchless: processing checkout
◇ 94fa741 3m (master) README
┣━┓
┃ ◯ f4e2e91 2m Something
┃ ┃
┃ ◯ e31bf9c 1m Update README
┃ ┃
┃ ◯ 4856148 1m Something else
┃
◯ 81816d2 0s Update README
┃
● 7118d31 0s Something else
In-memory rebase succeeded.
```

Perhaps your hooks are installed incorrectly (if you use `husky`, it may overwrite them aggressively), or your version of Git is too old. Could you reproduce the issue and upload the output of `git-branchless bug-report`? Also, after reproducing it in a worktree, run the same `bug-report` command in the main worktree as well and upload both, as I see now that they may report different hooks as having been installed. The `bug-report` subcommand will sanitize the commit graph and branch names, so you can try it on your real repo if you're comfortable with that.

**Reply by arxanas** [reply]
I was wrong; I think I never noticed the problems with worktrees in my previous usage. I opened https://github.com/arxanas/git-branchless/pull/1095 to address this.

> But now the "Update README" commit is gone from the smartlog!

The problem is that it was never reachable from the worktree event log to begin wit — iIt happened to be visible in the smartlog only because it was reachable by `HEAD`. Once we rebase and switch commits, we are no longer at the old `HEAD` and it's no longer visible in the worktree.
